Unbundle file engine binaries from plugin

jyscao opened the pull request:
Instead making users download 4 file engine binaries (3 of which they will never use or even all 4 if their OS+arch doesn't match any of them), it makes more sense to use a post install script to download only the binary applicable to their system.

jyscao committed:
Unbundle file engine binaries from plugin repo

* add post-install shell script to auto fetch file engine bin
* use MD5 to prevent unnecessary re-downloading
* update 'Installation' section of README
* remove existing bins from git tracking (don't delete bins for now)

jyscao commented:
Additional minor improvements:

- [ ] currently, the script gets triggered on all plugin updates even if the binaries themselves are unchanged. Proposed solution: add checksums for all fine engine binaries, then only re-download if checksums differ from user's existing binary in their vim-`ctrlspace/bin`

- [ ] add powershell/batch script to provide auto-download of `file_engine_windows_amd64.exe` for Windows users

Konfekt commented on bin/fetch-engine.sh:
```diff
@@ -0,0 +1,37 @@
+fi
+
+engine_fpath=$(dirname "$0")/${engine_name}
+curl -fSsLo "$engine_fpath" \
```
A good set of parameters to `curl`.
Preferably use their expanded synonyms; I think they stand for

```sh
curl --fail --show-error --remote-name --location
```

jyscao replied:
Using the long option names make sense, for people who want to inspect the script.

Though `--remote-name` which is long for `-O`, isn't actually what should be used here, b/c it then the downloaded file would contain a suffix of `?raw=true`

Konfekt commented on bin/fetch-engine.sh:
```diff
@@ -0,0 +1,37 @@
+  exit 1
+fi
+
+engine_fpath=$(dirname "$0")/${engine_name}
```
Since `dirname` can return spaces, double quotes deem necessary.

Konfekt reviewed (commented):
Very sensible improvement; thanks for the good work!

jyscao commented on README.md:
````diff
@@ -159,6 +179,11 @@ To find more about setting up the file engines, check:
 :help g:CtrlSpaceFileEngine
 ```
 
+__NOTE__: in a future major release of the plugin, the `g:CtrlSpaceFileEngine`
````
Yes, this is a very fair point. I will get rid of the deprecation warning for this addition.

jyscao replied:
I commented out this note for the time being.

We'll discuss this particular change more once I submit that issue (hopefully by this weekend) containing the alluded points, so you have a better context of where I'm coming from.

jyscao commented on autoload/ctrlspace/context.vim:
```diff
@@ -94,6 +94,12 @@ function! s:init() abort
 
     if s:conf.FileEngine ==# "auto"
         let s:conf.FileEngine = s:detectEngine()
+    else
```
I been meaning to submit an issue outlining a couple of my main medium to long-term plans/ideas for the project going forward, you'll know it when you see it.

Just as a preview though, one major point I want to raise is that, for an eventual major release we should remove some unnecessary options (which ones are deemed so is ofc up for discussion and should be agreed upon before the changes are committed).

jyscao commented:
@Konfekt I added a MD5 checksum check, and added some more details to the README. Lmk what you think; I think it's probably more or less good to be merged.

Although I'm not too sure if we should just remove those 4 binaries from tracking immediately. Since if that's done, next time people pull the update, it'll be deleted from their plugin, but if the script didn't work correctly, then they might get confused. But without removing the binaries from git tracking, then it defeats the whole point of this PR. What you think?

Konfekt commented:
Well, this change is abrupt, but for the better. The spell download vimscript file

https://github.com/vim/vim/blob/master/runtime/autoload/spellfile.vim

serves as a template to implement it without any need for vim-plug on every OS.
